src/Consistency.hs

```diff
@@ -0,0 +1,117 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE MonadComprehensions #-}
+{-# LANGUAGE ViewPatterns #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE DataKinds #-}
+{-# Language RecordWildCards #-}
+
+module Consistency (
+  checkCases
+) where
+
+
+import Prelude hiding (GT, LT)
+
+import Data.List
+import Text.PrettyPrint.ANSI.Leijen hiding ((<$>))
+import System.Exit (exitFailure)
+import System.IO (stdout)
+
+import Syntax
+import Syntax.Annotated
+import SMT
+import Syntax.Untyped (makeIface)
+
+-- TODO this is duplicated in hevm Keccak.hs but not exported
+combine :: [a] -> [(a,a)]
+combine lst = combine' lst []
+  where
+    combine' [] acc = concat acc
+    combine' (x:xs) acc =
+      let xcomb = [ (x, y) | y <- xs] in
+      combine' xs (xcomb:acc)
+
+-- | Checks non-overlapping cases
+mkNonoverlapAssertion :: [Exp ABoolean] -> Exp ABoolean
+mkNonoverlapAssertion caseconds =
+  mkOr $ (\(x, y) -> And nowhere x y) <$> combine caseconds
+  where 
+    mkOr [] = LitBool nowhere False
+    mkOr (c:cs) = Or nowhere c (mkOr cs)
+
+-- | Checks exhaustiveness of cases
+mkExhaustiveAssertion :: [Exp ABoolean] -> Exp ABoolean
+mkExhaustiveAssertion caseconds =
+  foldl mkAnd (LitBool nowhere True) caseconds
+  where
+    mkAnd r c = And nowhere (Neg nowhere c) r
+
+-- | Create a query for cases
+mkCaseQuery :: ([Exp ABoolean] -> Exp ABoolean) -> [Behaviour] -> (Id, SMTExp, (SolverInstance -> IO Model))
+mkCaseQuery props behvs@((Behaviour _ _ (Interface ifaceName decls) preconds _ _ _ _):_) =
+  (ifaceName, mkSMT, getModel)
+  where
+    locs = nub $ concatMap locsFromExp (preconds <> caseconds)
+    env = concatMap ethEnvFromBehaviour behvs
+    pres = mkAssert ifaceName <$> preconds
+    caseconds = concatMap _caseconditions behvs
+      
+    mkSMT = SMTExp
+      { _storage = concatMap (declareStorage [Pre]) locs
+      , _calldata = declareArg ifaceName <$> decls
+      , _environment = declareEthEnv <$> env
+      , _assertions = (mkAssert ifaceName $ props caseconds) : pres
+      }
+
+    getModel solver = do
+      prestate <- mapM (getStorageValue solver ifaceName Pre) locs
+      calldata <- mapM (getCalldataValue solver ifaceName) decls
+      environment <- mapM (getEnvironmentValue solver) env
+      pure $ Model
+        { _mprestate = prestate
+        , _mpoststate = []
+        , _mcalldata = (ifaceName, calldata)
+        , _menvironment = environment
+        , _minitargs = []
+        }
+mkCaseQuery _ [] = error "Internal error: behaviours cannot be empty"
+
+-- | Checks nonoverlapping and exhaustiveness of cases
+checkCases :: Act -> IO ()
+checkCases (Act _ contracts) = do
+  let groups = concatMap (\(Contract _ behvs) -> groupBy sameIface behvs) contracts
+  let config = SMT.SMTConfig CVC5 50000 True -- TODO make this parametrizable
```

Comment:
just tagging with a comment here to make sure it doesn't get forgotten before merge